### ISSUE TITLE
Rare 'Resource deadlock avoided' abort: a pool task can hold the last reference to its own ThreadPool, so ~ThreadPool joins the thread it runs on

### ISSUE THREAD

**Issue by mobile-threat-hunter** [mobile-threat-hunter/dex-analyzer-for-llm#50]
Observed **once**, and unreproduced — but the mechanism turned out to be structurally present in the vendored core, so this is filed on the code path rather than on the sighting.

## What was seen

During a routine test run (dexllm at `aa83f42`, native `.so` unchanged since v0.14.0):

```
$ .venv/bin/python -m pytest tests/test_doc_examples.py tests/test_capability_catalog.py \
                             tests/test_sdk.py tests/test_tools.py -q
  what():  Resource deadlock avoided
Fatal Python error: Aborted

Thread 0x00007daf32887780 (most recent call first):
  File ".../_pytest/capture.py", line 904 in pytest_runtest_teardown
```

**Reproduction attempts, all negative:** the same 4-file subset 11 more times, the full suite 3×, and a targeted loop of `summarize_capabilities` over the corpus (6 rounds × every APK). Only the tail of the output was captured, so the C++ frames are not available. Treat the sighting as weak evidence; the code path below is the strong part.

## The mechanism, traced in the tree

`what() == "Resource deadlock avoided"` is `std::errc::resource_deadlock_would_occur`. The dominant producer is **`std::thread::join()` called on the current thread**, and that is reachable here:

1. [`ThreadPool::~ThreadPool()`](../blob/master/vendor/dexkit_core/Core/third_party/thread_helper/ThreadPool.h#L143) joins every worker unconditionally:
   ```cpp
   for (std::thread &worker: workers)
       worker.join();
   ```
2. [`QueryScheduler::EnqueueDispatchTasks`](../blob/master/vendor/dexkit_core/Core/dexkit/include/query_scheduler.h#L569) captures a **`shared_ptr` to itself** into the lambda that runs *on that pool*:
   ```cpp
   auto self = shared_from_this();
   pool_->enqueue([self, dispatch_task = std::move(dispatch_task)]() mutable {
       TaskCompletionGuard completion_guard(self, dispatch_task.query_id);
       dispatch_task.task();
   });
   ```
   (`TaskCompletionGuard` holds another `shared_ptr<QueryScheduler>` for the duration.)
3. `QueryScheduler` owns `pool_` as a `shared_ptr<ThreadPool>`.
4. `DexKit` drops both in [`SetThreadNum`](../blob/master/vendor/dexkit_core/Core/dexkit/dexkit.cpp#L113) and at destruction (`shared_query_scheduler_.reset(); shared_query_pool_.reset();`).

So if the owning `DexKit` releases its references while a dispatched lambda is still alive on a worker, the **last** `shared_ptr<QueryScheduler>` is the one captured in that lambda. Destroying the lambda on the worker destroys the scheduler there, which releases the last `shared_ptr<ThreadPool>`, so `~ThreadPool()` runs **on one of its own workers** and joins the thread it is executing on → `std::system_error` → uncaught in a thread → `std::terminate` → the observed abort.

That fits every detail of the sighting: it is a **destruction-time race** (hence rare and order-dependent), and it landed at `pytest_runtest_teardown`, which is exactly where a `DexKit` fixture is dropped and the interpreter collects it.

A second, weaker candidate: libstdc++'s `std::shared_mutex::lock()` also throws this on a same-thread write re-lock (`Decompiler::cache_mutex_`, [decompiler.cpp](../blob/master/native/dad_cpp/decompiler.cpp#L123)). No re-entrant path was found there by inspection, and it would not explain a teardown-time abort as naturally.

## Why it matters beyond the test suite

An abort is uncatchable — `safe.py`'s wrapper guards hangs, not crashes, and the whole point of the load-time verifier work is that dexllm does not die on analysis. Any consumer that finishes with a `DexKit` while a query is in flight is in the same position; a long-lived server (`server.py` sessions are LRU-evicted, so a `DexKit` is dropped while requests may still be running) is the realistic exposure.

## Suggested direction — not yet decided

- **Break the cycle at the capture**: the task needs the scheduler only to report completion, so a `weak_ptr` promoted inside the task would keep the last reference off the worker. Changes an upstream-vendored file.
- **Or make `~ThreadPool` self-join-safe**: `if (worker.get_id() != std::this_thread::get_id()) worker.join(); else worker.detach();` — smaller, but leaves the pool being destroyed from inside itself, which has its own hazards.
- **Or keep the pool alive past the scheduler** so the last release cannot land on a worker.

Whichever, it needs a **deterministic reproduction first** — most cheaply a targeted C++ or Python test that drops the last `DexKit` reference while a query is dispatched (e.g. a large `find_*` on a many-dex source, released from another thread mid-query), run under a loop. Without that, a fix cannot be shown to work.

## Scope note

The code is in `vendor/dexkit_core/Core` (upstream DexKit), which dexllm ships and already patches. If the fix lands here it belongs in the vendored-patch set; it may also be worth reporting upstream.

**Comment by mobile-threat-hunter** [comment]
## Update: the premise was wrong — this reproduces from the plain Python API, in under a second

The issue was filed on the code path because the sighting looked unreproducible ("Treat the sighting as weak evidence"). That framing no longer holds. **The mechanism is reachable from `dexllm`'s own Python API, single-threaded, with no concurrency and no load**, and it takes a few hundred iterations.

```python
import dexllm
while True:                                             # test_apk/APK/Test.dex — 552 B, 1 class
    dk = dexllm.DexKit("test_apk/APK/Test.dex")
    dk.find_classes_by_name("a", match_type="contains")
    del dk
```

| | result |
|---|---|
| single process, ×5 runs | **5/5 abort**, at iteration **0 / 400 / 500 / 700 / 850** |
| 48 processes in parallel | **48/48 abort**, every one with `what(): Resource deadlock avoided` |

The loop runs at ~790 iter/s on this source, so a crash arrives in roughly a second of looping.

### Why the original hammer missed it: the source SIZE is the variable

Same script, same machine, larger sources:

| source | classes | rate | result |
|---|---|---|---|
| `Test.dex` | 1 | 792/s | dies within ~500 iterations (5/5) |
| `hello-world.apk` | 2,119 | 55/s | **0 crashes** in 2×100 s (≈5,500 iterations each) |
| `a2dp.Vol_137.apk` | 1,353 | — | **0 crashes** in 4,800 iterations across 16 processes |

The reason is `~DexKit`'s member order. `shared_query_pool_` is declared *before* `shared_query_scheduler_`, and `~DexKit` is `= default`, so the pool reference is released **last** — after `dex_items` and `images` have been destroyed. Crashing requires the main thread to release both references *before* the worker reaches the end of its loop iteration, and on a large APK that teardown costs milliseconds while the worker only needs a few atomic decrements. On a 1-class dex the teardown is free, so the main thread wins routinely.

**The original reproduction attempt used a 1,353-class APK — i.e. close to the most survival-favourable condition available.** That is the whole reason it read as "unreproducible".

### The window, precisely

One iteration of the worker loop ([`ThreadPool.h`](../blob/master/vendor/dexkit_core/Core/third_party/thread_helper/ThreadPool.h)):

```cpp
for (;;) {
    std::function<void()> task;                    // (1)
    { lock; wait; if (stop && tasks.empty()) return; task = std::move(front); pop; }
    task();                                        // (2)
}                                                  // (3)  <- dies here
```

`task` transitively owns `shared_ptr<packaged_task>` → the dispatch lambda → its captured `self` (`QueryScheduler`) → `pool_` (`ThreadPool`).

* Inside **(2)**, `dispatch_task.task()` runs the search closure — **the caller's future resolves here** — then `~TaskCompletionGuard` → `OnTaskFinished` releases *its* copy of `self` (never the last one; the lambda still holds one).
* At **(3)** the iteration ends, the local `std::function` dies, the lambda is destroyed, and its `self` is released.

If, by then, `DexKit` has already dropped both `shared_query_scheduler_` and `shared_query_pool_` and no other query is live, that release is the **last** one:

```
~QueryScheduler   (on a worker)
  └ pool_ released → last
      └ ~ThreadPool   (on a worker)
          └ workers[i].join()  where workers[i] IS the current thread
              → std::system_error escapes a destructor → std::terminate
```

So the window opens the moment the caller's future resolves and closes at (3). Everything the main thread does after the query returns — `~SharedThreadPoolQueryExecutor`/`DetachQuery`, result marshalling, the return to Python, `del dk`, `~DexKit` — has to fit inside it.

### Deterministic C++ reproduction

Mirroring `DexKit`'s ownership exactly (owner releases scheduler then pool while a task is in flight), using the real `ThreadPool` + `QueryScheduler`: **8/8 abort**, and gdb pins the frame:

```
#14 ThreadPool::~ThreadPool ()      at .../thread_helper/ThreadPool.h:151     <- worker.join()
#18 std::_Sp_counted_ptr_inplace<ThreadPool, ...>::_M_dispose ()
#21 std::_Sp_counted_base<...>::_M_release ()
... running on Thread 2, i.e. one of that pool's own workers
```

### Upstream status

The whole defect surface is upstream `LuckyPray/DexKit`, unmodified by this repo:

* `Core/dexkit/include/query_scheduler.h` and `Core/dexkit/include/query_executor.h` on upstream **master** are **byte-identical** to our vendored copies; `ThreadPool.h` differs only by our WASM `#if` shim (23 lines, compiled out on native).
* Our tree has never edited any of them beyond the relocate commit (`93f192a`) and the WASM shim (`7722f34`).
* The ownership was introduced by upstream **`5226b737` (2026-04-11) "feat(query): introduce native query scheduler and shared-pool execution"** — the commit that added `shared_query_pool_` / `shared_query_scheduler_` and the scheduler itself. It touched `ThreadPool.h` too, but not the destructor: the unconditional join descends from the 2012 progschj pool, where it was *correct* because a pool was a scope-owned local destroyed by its creating thread. That commit is where the pool became `shared_ptr`-owned by an object reachable from a worker, and the implicit precondition silently stopped holding.
* No prior upstream report: searches for `ThreadPool` / `deadlock` / `join` return nothing; the one crash issue (#22, 2023) is an unrelated `nativeInitDexKit` ANR.

Upstream consumers are **more** exposed than dexllm: the search bindings here hold the GIL (so a query and a destruction cannot overlap, leaving only the tail window), and dexllm never sets `find_first` — whereas early exit makes `DetachQuery` dispatch leftover tasks from the *calling* thread, which run on workers long after the query returned and widen the window enormously.

### What a fix has to satisfy — two attempts, both refuted with running reproductions

Recorded so the next attempt doesn't repeat them.

1. **Drain before releasing** (`ThreadPool::Shutdown()` called from `~DexKit` / `SetThreadNum` / the pool-replacement branch). **Rejected — it is a net regression.** `enqueue` has `if (stop) abort();`, `Shutdown()` sets `stop` and *then* joins, and the join waits for running tasks whose tail (`OnTaskFinished` → `EnqueueDispatchTasks` → `enqueue`) fires after `stop` is set. Any query with more tasks than workers — i.e. the normal shape — aborts. Measured: pre-change 3/3 complete cleanly, with the drain **3/3 SIGABRT**, backtrace through `ThreadPool::enqueue` on a worker.
2. **`QueryScheduler::pool_` → `weak_ptr`, alone.** Closes the lambda's `self` (verified: reverting it makes the guard abort 3/3) but **not** the bug: `EnqueueDispatchTasks` then takes a transient `pool_.lock()` which also runs on a worker via `OnTaskFinished`, and that transient can be the last reference. Measured with 400 tasks/query: **aborts on 5/5 runs**, same `Resource deadlock avoided`, backtrace at the `lock()`'s destruction. A query-lifetime keepalive on `SharedThreadPoolQueryExecutor` narrows it but does not remove it — it only holds while the executor lives.
   * Its expired-branch also must not simply drop the tasks: a dispatch task backs a `std::future` somebody is waiting on, so dropping one gives `broken_promise` out of `FindClass`/`FindMethod`/`FindField` (which have no `try`/`catch` anywhere) or an outright permanent hang, plus a permanent `in_flight` leak. Measured: 2 futures `broken_promise`, 1 never ready.

So the constraint is: **a worker must not hold a strong pool reference at all** — which in turn requires `enqueue`'s `abort()` on a stopped pool to go first (a stopped pool still drains its queue, so an enqueue after `stop` would run) — **or** `~ThreadPool` itself has to be safe when it runs on one of its own workers, which means the pool's state has to outlive the object (the worker loop and `enqueue` both capture `this` today).

### Severity

Suggest upgrading from "seen once, unreproduced". It is an uncatchable process abort (`safe.py` guards hangs, not crashes) that lands at **destruction time**, so a completed analysis is lost along with the process, and the Python traceback points at wherever the main thread happened to be — which is why the original sighting looked like a `pytest`/`capture.py` problem. The workload that hits it is ordinary: open a small dex, search it, drop it, repeat — bulk triage over extracted/dumped dexes, `add_dumped_dexes` flows, per-file scanning loops, and test fixtures.

<sub>Measured on Linux x86-64, 32 cores, libstdc++ 14, Python 3.13, dexllm `37c68d2` (v0.15.0) with the unmodified vendored core.</sub>

**Comment by mobile-threat-hunter** [comment]
Fixed in `2e515ca`.

The issue asked for a **deterministic reproduction first** — "without that, a fix cannot be shown to work". There is one: a 30-line program in which a task owns its own pool aborts **10/10** with the exact reported `Resource deadlock avoided`, and so does the scheduler-shaped variant (task → intermediate owner → pool). Both are now committed as guard cases.

## The fix

Everything a worker touches moved into a `struct State` held by `shared_ptr`, so it **outlives the object**; the worker lambda and the packaged task capture `state` rather than `this` (a queued task can outlive the pool, so reading `should_skip_task` through `this` was a latent use-after-free of its own); and the destructor **detaches the ONE worker it is running on** while joining the rest. Destruction from a non-worker is unchanged — it still joins every worker, which is the contract the rest of the core relies on.

**Why not the other two directions the issue listed.** A `weak_ptr` at the dispatch is not sufficient: `TaskCompletionGuard` must hold a STRONG reference while it reports completion, and any strong reference on a worker can become the last one. Keeping the pool alive past the scheduler does not help either, because the scheduler owns the pool, so a scheduler destroyed on a worker takes the pool with it wherever the owner's own references went. Fixing it inside `ThreadPool` covers every present and future ownership mistake instead of one call site.

## Two more defects the reviews found in the fix itself

- **the detach opened a SECOND use-after-free.** `ReleaseMatcherThreadLocalCaches` used to run once every worker was dead; with one detached, the same call deleted the matcher cache of a thread that is still RUNNING — and `dex_item_matcher.cpp` holds it in a `thread_local` **raw** pointer that no other thread can reset, so the detached worker would dereference freed memory on its next task (reproduced as an ASan `heap-use-after-free`). Only the ids of threads that were actually JOINED are released now; the detached worker keeps its cache — one leaked cache on a teardown-only path instead of a UAF. `thread_ids` left `State` with that change: the destructor reads the ids off the `std::thread` objects, which is race-free and already correct before a worker has started.
- **the constructor was not exception-safe.** `std::thread`'s ctor throws when the process is out of threads, and unwinding a partially built pool destroys a still-joinable `std::thread` — `std::terminate`. Reproduced with a real `RLIMIT_NPROC`; failing the FIRST thread was survivable, the SECOND aborted. (This one also silently voided the recovery path dexllm#55 builds on that scope being catchable, which is why the two were fixed together.)

## Accepted costs, stated rather than discovered

The self-destruct path leaves one **detached** thread where before there were none, so an `exit()` in that window can run it during static destruction. And on that path the destructor no longer implies "every queued task has finished" — the detached worker keeps draining AFTER it returns and after the owner is gone. Only a single-worker pool can reach it (any other worker drains before its join returns) and DexKit cannot, because a queued dispatch lambda holds a reference that would have kept the pool alive. The queue is deliberately NOT cleared instead: dropping a queued task leaves its `std::future` unsatisfied forever, which is the failure mode dexllm#55 exists to remove.

## Verification

a/b OFF vs ON (same script, both `.so` md5-verified): **406 records over 31 loadable sources x 13 axes, identical sha256, 0 diff**. ASan and TSan on the reproducer clean, 5 runs each; an adversarial reviewer additionally ran 1800 self-destruct rounds under both with 0 warnings. parity 29/29, pytest 631, determinism 3 seeds identical, perf within noise.

Guards in `tests/parity/thread_pool_selfdestruct_test.cpp` (7 cases). It **stubs** the matcher-cache registry rather than linking `dexkit_static`, so a case can assert WHICH ids the destructor released — without that, the live-thread fix survived a reviewer's mutant **15/15**. Six mutants killed 5/5 each: the pre-fix header, the worker capturing `this`, the packaged task capturing `this`, detach-all, release-the-live-id, and drop-the-release.

**Two of the guards were worthless until they were mutated**, which is the part worth recording. The first cut observed the wrong MOMENT — a flag set by the task fires BEFORE the pool is destroyed, so case 2 passed against the pre-fix header; every case now signals from a `shared_ptr` deleter, which runs strictly after `~ThreadPool` returns. And every self-destruct case RACED: if the task finished before the enqueuing thread dropped its reference, the destruction happened on the MAIN thread and the case passed having exercised nothing — so each now parks its task until the enqueuing thread has released, and ASSERTS which thread ran the destructor.
